### DISCUSSION TITLE
Some methods on IMetadataDictionary will flag the original document as modified

### DISCUSSION THREAD

**Discussion by stevehansen** [ravendb/ravendb#13604] [Q&A]
Some context as to our exact scenario that caused this issue:
- we are using subscriptions (checking whether ```@refresh``` does not exists)
- we have a IDocumentStore.OnBeforeStore hook to set a custom ModifiedOn property

We ran into this issue when our subscription wants to remove the refresh metadata, using

```c#
var metadataFor = session.Advanced.GetMetadataFor(entity);
// if no longer needed
metadataFor.Remove(Constants.Documents.Metadata.Refresh);
```

Looking at the existing code:
- https://github.com/ravendb/ravendb/blob/v5.3/src/Raven.Client/Json/MetadataAsDictionary.cs#L187-L217
- https://github.com/ravendb/ravendb/blob/v5.3/src/Raven.Client/Json/MetadataAsDictionary.cs#L51

These method calls will force a call to Initialize, which will always flag the document as dirty. Which caused our subscription to get in an infinite loop (and to make matters worse we had revision enabled which exploded the database with revisions).

Remove/Clear can be easily worked around (we do a ```metadataFor.ContainsKey(Constants.Documents.Metadata.Refresh)``` now before calling Remove, you could use ```metadataFor.Count > 0``` for Clear).

For GetEnumerator/CopyTo this would be impossible, those will always force the entity to be flagged as dirty.

**Comment by ppekrol** [comment]
I moved this to our bugtracker. Link: https://issues.hibernatingrhinos.com/issue/RavenDB-17996
